robot-server/robot_server/protocols/analysis_store.py

```
@@ -64,31 +98,263 @@ def update(
+            except sqlalchemy.exc.NoResultFound:
+                raise AnalysisNotFoundError(analysis_id=analysis_id)
+        return _CompletedAnalysisResource.from_sql_row(result)
+
+    def get_by_protocol(self, protocol_id: str) -> List[_CompletedAnalysisResource]:
+        statement = (
+            sqlalchemy.select(analysis_table)
+            .where(analysis_table.c.protocol_id == protocol_id)
+            .order_by(sqlite_rowid)
+        )
+        with self._sql_engine.begin() as transaction:
+            self._check_protocol_exists(connection=transaction, protocol_id=protocol_id)
+            results = transaction.execute(statement).all()
+        return [_CompletedAnalysisResource.from_sql_row(r) for r in results]
+
+    def get_ids_by_protocol(self, protocol_id: str) -> List[str]:
+        statement = (
+            sqlalchemy.select(analysis_table.c.id)
+            .where(analysis_table.c.protocol_id == protocol_id)
+            .order_by(sqlite_rowid)
+        )
+        with self._sql_engine.begin() as transaction:
+            self._check_protocol_exists(connection=transaction, protocol_id=protocol_id)
+            results = transaction.execute(statement).all()
+
+        result_ids: List[str] = []
+        for row in results:
+            assert isinstance(row.id, str)
+            result_ids.append(row.id)
+
+        return result_ids
+
+    def add(self, completed_analysis_resource: _CompletedAnalysisResource) -> None:
+        statement = analysis_table.insert().values(
+            completed_analysis_resource.to_sql_values()
+        )
+        with self._sql_engine.begin() as transaction:
+            transaction.execute(statement)
+
+    def check_protocol_exists(self, protocol_id: str) -> None:
+        with self._sql_engine.begin() as transaction:
+            self._check_protocol_exists(connection=transaction, protocol_id=protocol_id)
+
+    @staticmethod
+    def _check_protocol_exists(
```

Why do we need this? inserting with a FK that does not exist should raise an error.

---

For the case where we're adding a *pending* analysis. Since that's in-memory only, it doesn't involve adding a SQL row, so SQL can't enforce foreign key constraints.

---

Discussed in call: we've decided `AnalysisStore` isn't responsible for catching this error, so this can be removed.